File: 3.x.x/crates-utopia/dns/src/resolver/memory.rs
use crate::error::Result;
use crate::query::Query;
use crate::resolver::Resolver;
use crate::zone::{self, Zone};
use crate::Message;

/// In-memory authoritative resolver. PHP `Utopia\DNS\Resolver\Memory`.
#[derive(Debug, Clone)]
pub struct Memory {
    zone: Zone,
}

impl Memory {
    #[must_use]
    pub fn new(zone: Zone) -> Self {
        Self { zone }
    }
}

impl Resolver for Memory {
    fn resolve(&self, query: &Query) -> Result<Message> {
        zone::Resolver::lookup(&query.message, &self.zone)
    }
}

File: packages/dns/src/DNS/Resolver/Memory.php
<?php

declare(strict_types=1);

namespace Utopia\DNS\Resolver;

use Utopia\DNS\Message;
use Utopia\DNS\Query;
use Utopia\DNS\Resolver;
use Utopia\DNS\Zone;
use Utopia\DNS\Zone\Resolver as ZoneResolver;

class Memory implements Resolver
{
    public function __construct(private readonly Zone $zone) {}

    public function resolve(Query $query): Message
    {
        return ZoneResolver::lookup($query->message, $this->zone);
    }
}
